service/worker/esanalyzer/workflow.go

```diff
@@ -0,0 +1,389 @@
+		ScheduleToStartTimeout: time.Minute,
+		StartToCloseTimeout:    3 * time.Minute,
+		RetryPolicy:            &retryPolicy,
+	}
+	refreshStuckWorkflowsOptions = workflow.ActivityOptions{
+		ScheduleToStartTimeout: time.Minute,
+		StartToCloseTimeout:    10 * time.Minute,
+		RetryPolicy:            &retryPolicy,
+	}
+
+	wfOptions = cclient.StartWorkflowOptions{
+		ID:                           esAnalyzerWFID,
+		TaskList:                     taskListName,
+		ExecutionStartToCloseTimeout: 5 * 24 * time.Hour,
+		CronSchedule:                 "0 * * * *", // "At minute 0" => every hour
+	}
+)
+
+func init() {
+	workflow.RegisterWithOptions(Workflow, workflow.RegisterOptions{Name: esanalyzerWFTypeName})
+	activity.RegisterWithOptions(GetWorkflowTypes, activity.RegisterOptions{Name: getWorkflowTypesActivity})
+	activity.RegisterWithOptions(FindStuckWorkflows, activity.RegisterOptions{Name: findStuckWorkflowsActivity})
+	activity.RegisterWithOptions(
+		RefreshStuckWorkflowsFromSameWorkflowType,
+		activity.RegisterOptions{Name: refreshStuckWorkflowsActivity},
+	)
+}
+
+// Workflow queries ElasticSearch to detect issues and mitigates them
+func Workflow(ctx workflow.Context) error {
+	// list of workflows with avg workflow duration
+	var wfTypes []WorkflowTypeInfo
+	err := workflow.ExecuteActivity(
+		workflow.WithActivityOptions(ctx, getWorkflowTypesOptions),
+		getWorkflowTypesActivity,
+	).Get(ctx, &wfTypes)
+	if err != nil {
+		return err
+	}
+
+	for _, info := range wfTypes {
+		var stuckWorkflows []*WorkflowInfo
+		err := workflow.ExecuteActivity(
+			workflow.WithActivityOptions(ctx, findStuckWorkflowsOptions),
+			findStuckWorkflowsActivity,
+			info,
+		).Get(ctx, &stuckWorkflows)
+		if err != nil {
+			return err
+		}
+		if len(stuckWorkflows) == 0 {
+			continue
+		}
+
+		err = workflow.ExecuteActivity(
+			workflow.WithActivityOptions(ctx, refreshStuckWorkflowsOptions),
+			refreshStuckWorkflowsActivity,
+			stuckWorkflows,
+		).Get(ctx, nil)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// RefreshStuckWorkflowsFromSameWorkflowType is activity to refresh stuck workflows from the same domain
+func RefreshStuckWorkflowsFromSameWorkflowType(
+	ctx context.Context,
+	workflows []WorkflowInfo,
+) error {
+	analyzer := ctx.Value(analyzerContextKey).(*Analyzer)
+	domainID := workflows[0].DomainID
+	domainEntry, err := analyzer.domainCache.GetDomainByID(domainID)
+	if err != nil {
+		analyzer.logger.Error("Failed to get domain entry",
+			tag.WorkflowDomainID(domainID),
+		)
+		return err
+	}
+	domainName := domainEntry.GetInfo().Name
+	clusterName := domainEntry.GetReplicationConfig().ActiveClusterName
+
+	adminClient := analyzer.clientBean.GetRemoteAdminClient(clusterName)
+	for _, workflow := range workflows {
+		if workflow.DomainID != domainID {
+			return types.InternalServiceError{
+				Message: fmt.Sprintf(
+					"Inconsistent worklow. Expected domainID: %v, actual: %v",
+					domainID,
+					workflow.DomainID),
+			}
+		}
```

Comment:
ah.  yeah, I'd assume that will happen eventually, though most likely won't.
we could probably check our metrics and discover how many there are lately fwiw.

Reply:
welp.  tons of collisions, though most are similar enough that it might not skew results.
still, probably a sign that it's more common than we thought, may be worth addressing up-front.